rocAL/source/loaders/audio/audio_loader_sharded.cpp

```diff
@@ -0,0 +1,166 @@
+        increment_loader_idx();
+}
+
+LoaderModuleStatus AudioLoaderSharded::load_next() {
+    if(!_initialized)
+        return LoaderModuleStatus::NOT_INITIALIZED;
+    increment_loader_idx();
+    // Since loaders may have different number of audios loaded, some run out earlier than other.
+    // Fast forward through loaders that are empty to get to a loader that is not empty.
+    fast_forward_through_empty_loaders();
+    auto ret = _loaders[_loader_idx]->load_next();
+    return ret;
+}
+
+void
+AudioLoaderSharded::initialize(ReaderConfig reader_cfg, DecoderConfig decoder_cfg, RocalMemType mem_type,
+                               unsigned batch_size, bool keep_orig_size) {
+    if(_initialized)
+        return;
+    _shard_count = reader_cfg.get_shard_count();
+    // Create loader modules
+    for(size_t i = 0; i < _shard_count; i++) {
+        std::shared_ptr loader = std::make_shared<AudioLoader>(_dev_resources);
+        loader->set_prefetch_queue_depth(_prefetch_queue_depth);
+        _loaders.push_back(loader);
+    }
+    // Initialize loader modules
+    for(size_t idx = 0; idx < _shard_count; idx++) {
+        _loaders[idx]->set_output(_output_tensor);
+        _loaders[idx]->set_gpu_device_id(idx);
+        reader_cfg.set_shard_count(_shard_count);
+        reader_cfg.set_shard_id(idx);
+        _loaders[idx]->initialize(reader_cfg, decoder_cfg, mem_type, batch_size, keep_orig_size);
+    }
+    _initialized = true;
+}
+
+void AudioLoaderSharded::start_loading() {
+    for(unsigned i = 0; i < _loaders.size(); i++) {
+        _loaders[i]->start_loading();
+    //  Changing thread scheduling policy and it's priority does not help on latest Ubuntu builds
+    //  and needs tweaking the Linux security settings , can be turned on for experimentation
+#if 0
+        // Set thread scheduling policy
+        struct sched_param params;
+        params.sched_priority = sched_get_priority_max(SCHED_FIFO);
+        _loaders[i]->set_cpu_sched_policy(params);
+#endif
+        // Setting cpu affinity for threads works and can be activated below for experimentation
+#if 0
+        // Set thread affinity thread 0 to core 0 , 1 toc core 1 , ...
+        cpu_set_t cpuset;
+        CPU_ZERO(&cpuset);
+        CPU_SET(i, &cpuset);
+        _loaders[i]->set_cpu_affinity(cpuset);
+#endif
+    }
+}
+
+void AudioLoaderSharded::shut_down() {
+    for(unsigned i = 0; i < _loaders.size(); i++)
+        _loaders[i]->shut_down();
+}
+
+void AudioLoaderSharded::set_output(Tensor* output_tensor) {
+    _output_tensor = output_tensor;
+}
+
+size_t AudioLoaderSharded::remaining_count() {
+    int sum = 0;
+    for(auto& loader: _loaders)
+        sum += loader->remaining_count();
+    return sum;
+}
+
+void AudioLoaderSharded::reset() {
+    for(auto& loader: _loaders) {
+        loader->reset();
+        // loader->last_batch_padded_size();
+    }
+}
+
+void AudioLoaderSharded::increment_loader_idx() {
+    _loader_idx = (_loader_idx + 1)%_shard_count;
```

Comment:
Please add space before and after % operator

Reply:
handled in the formatting